packages/assets-controllers/src/balances.ts

```diff
@@ -0,0 +1,800 @@
+import type { AccountTreeControllerState } from '@metamask/account-tree-controller';
+import type { AccountWalletObject } from '@metamask/account-tree-controller';
+import type { AccountGroupObject } from '@metamask/account-tree-controller';
+import type { AccountsControllerState } from '@metamask/accounts-controller';
+import { isEvmAccountType } from '@metamask/keyring-api';
+import type { EntropySourceId } from '@metamask/keyring-api';
+import type { InternalAccount } from '@metamask/keyring-internal-api';
+import type { Hex } from '@metamask/utils';
+import type { CaipAssetType, CaipChainId } from '@metamask/utils';
+import {
+  KnownCaipNamespace,
+  parseCaipAssetType,
+  parseCaipChainId,
+  isStrictHexString,
+} from '@metamask/utils';
+
+import type { CurrencyRateState } from './CurrencyRateController';
+import type { MultichainAssetsRatesControllerState } from './MultichainAssetsRatesController';
+import type { MultichainBalancesControllerState } from './MultichainBalancesController';
+import type { TokenBalancesControllerState } from './TokenBalancesController';
+import type { TokenRatesControllerState } from './TokenRatesController';
+import type { TokensControllerState } from './TokensController';
+
+export type AccountGroupBalance = {
+  walletId: string;
+  groupId: string;
+  totalBalanceInUserCurrency: number;
+  userCurrency: string;
+};
+
+export type WalletBalance = {
+  walletId: string;
+  groups: Record<string, AccountGroupBalance>;
+  totalBalanceInUserCurrency: number;
+  userCurrency: string;
+};
+
+export type AllWalletsBalance = {
+  wallets: Record<string, WalletBalance>;
+  totalBalanceInUserCurrency: number;
+  userCurrency: string;
+};
+
+export type BalanceChangePeriod = '1d' | '7d' | '30d';
+
+export type BalanceChangeResult = {
+  period: BalanceChangePeriod;
+  currentTotalInUserCurrency: number;
+  previousTotalInUserCurrency: number;
+  amountChangeInUserCurrency: number;
+  percentChange: number;
+  userCurrency: string;
+};
+
+const isChainEnabledByMap = (
+  map: Record<string, Record<string, boolean>> | undefined,
+  id: Hex | CaipChainId,
+): boolean => {
+  if (!map) {
+    return true;
+  }
+  const isHex = isStrictHexString(id as string);
+  if (isHex) {
+    const evm = map[String(KnownCaipNamespace.Eip155)];
+    return Boolean(evm?.[id as Hex]);
+  }
+  const { namespace } = parseCaipChainId(id as CaipChainId);
+  return Boolean(map[namespace]?.[id as CaipChainId]);
+};
+
+const getWalletIdFromGroupId = (groupId: string): EntropySourceId => {
+  return groupId.split('/')[0] as EntropySourceId;
+};
+
+const getInternalAccountsForGroup = (
+  accountTreeState: AccountTreeControllerState,
+  accountsState: AccountsControllerState,
+  groupId: string,
+): InternalAccount[] => {
+  const walletId = getWalletIdFromGroupId(groupId);
+  const wallet = (
+    accountTreeState.accountTree.wallets as Record<string, AccountWalletObject>
+  )[walletId];
```

Comment:
Remove assertion, not needed now.

```suggestion
  const wallet = accountTreeState.accountTree.wallets[walletId];
```